File: encoding/bench/bench.c
// UTF-8 through the locale converters. C has no standard hex codec, so hex is left out.
#include <locale.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <wchar.h>

#define REPS (1 << 18)

static const wchar_t SEED[] = L"Hello \u00e9 \u03a9 \u20ac \u4e2d \U0001f600\n";

static double now_ms(void) {
  struct timespec t;
  clock_gettime(CLOCK_MONOTONIC, &t);
  return t.tv_sec * 1e3 + t.tv_nsec / 1e6;
}

static uint32_t chk_w(const wchar_t *s, size_t n) {
  uint32_t h = 0;
  for (size_t i = 0; i < n; i++) h = h * 31 + (uint32_t)s[i];
  return h + (uint32_t)n;
}

static uint32_t chk_b(const unsigned char *s, size_t n) {
  uint32_t h = 0;
  for (size_t i = 0; i < n; i++) h = h * 31 + s[i];
  return h + (uint32_t)n;
}

int main(void) {
  if (!setlocale(LC_CTYPE, "C.UTF-8") && !setlocale(LC_CTYPE, "en_US.UTF-8")) return 1;
  size_t seed = wcslen(SEED), cps = seed * REPS;
  wchar_t *text = malloc((cps + 1) * sizeof *text);
  for (size_t r = 0; r < REPS; r++) wmemcpy(text + r * seed, SEED, seed);
  text[cps] = 0;

  char *octets = malloc(cps * 4 + 1);
  const wchar_t *src = text;
  mbstate_t st = {0};
  double t0 = now_ms();
  size_t nb = wcsrtombs(octets, &src, cps * 4 + 1, &st);
  double ms = now_ms() - t0;
  if (nb == (size_t)-1) return 1;
  printf("utf8_encode\t%.3f\t%u\n", ms, chk_b((unsigned char *)octets, nb));

  wchar_t *back = malloc((cps + 1) * sizeof *back);
  const char *bsrc = octets;
  st = (mbstate_t){0};
  t0 = now_ms();
  size_t nw = mbsrtowcs(back, &bsrc, cps + 1, &st);
  ms = now_ms() - t0;
  if (nw == (size_t)-1) return 1;
  printf("utf8_decode\t%.3f\t%u\n", ms, chk_w(back, nw));
  return 0;
}

File: encoding/bench/bench.rs
// Rust has no standard hex codec, so hex is left out.
use std::time::Instant;

const REPS: usize = 1 << 18;

fn chk(xs: impl Iterator<Item = u32>) -> u32 {
    let (mut h, mut n) = (0u32, 0u32);
    for x in xs {
        h = h.wrapping_mul(31).wrapping_add(x);
        n = n.wrapping_add(1);
    }
    h.wrapping_add(n)
}

fn main() {
    let text: Vec<char> = "Hello \u{e9} \u{3a9} \u{20ac} \u{4e2d} \u{1f600}\n".repeat(REPS).chars().collect();

    let t0 = Instant::now();
    let octets: String = text.iter().collect();
    let ms = t0.elapsed().as_secs_f64() * 1e3;
    println!("utf8_encode\t{:.3}\t{}", ms, chk(octets.bytes().map(u32::from)));

    let bytes = octets.into_bytes();
    let t0 = Instant::now();
    let back: Vec<char> = String::from_utf8_lossy(&bytes).chars().collect();
    let ms = t0.elapsed().as_secs_f64() * 1e3;
    println!("utf8_decode\t{:.3}\t{}", ms, chk(back.iter().map(|&c| c as u32)));
}
